### DISCUSSION TITLE
Complex JS Backend stuff

### DISCUSSION THREAD

**Discussion by TheOneValen** [TriliumNext/Trilium#3436] [Q&A] [answered]
Hi!

I would like do do something more complicated in the backend using the custom request handler (I would like to mimic a bookmark sync api, so I can sync bookmarks to trilium). There are a few libraries I have to use to make it feasable. I am not a JS expert (I did some Typescript).

I tried to put a library (lzutf8) into a backend script js note. Then in the parent node I tried to include it with `require`. Is this how it supposed to work? I do not get an error on the require but the lib does not seem to work. Sorry if I miss some basic stuff here....

Thanks!

**Comment by zadam** [comment, marked as answer]
> Then in the parent node I tried to include it with require. Is this how it supposed to work? I do not get an error on the require but the lib does not seem to work. Sorry if I miss some basic stuff here....

Hi, yes, it's supposed to work like that. Trilium tries to emulate Common JS modules, but I think that not all libraries will work.

It's best to try to avoid dependencies like that.

**Reply by TheOneValen** [reply]
Thanks! My POC for decrypting the bookmarks (I try to re-implement xbrowsersync) worked by putting the lib into the parent note. Resulting in a very big note (just for some tiny part of the implementation). Big notes seem to load quite slowly, is that a known issue? What is the best practice here? Code in another app c&p into trilium? How can I debug my script? Any way I can write into the backend log?

**Reply by TheOneValen** [reply]
Yes, very curius, embedding [it ](https://unpkg.com/lzutf8@0.6.3/build/production/lzutf8.js) into the script works, require on a "sub" script does not.

**Reply by zadam** [reply]
Thanks for the library, I checked how it detects common.js and fixed the responsible code in Trilium so this now works (on master):

```js
var lz = require('lzutf8');

var output = lz.compress("test");
```

> Big notes seem to load quite slowly, is that a known issue?

Yes, but there's not much which can be done about it.

> What is the best practice here? Code in another app c&p into trilium?

The best practice it keep the notes short, but that's not helpful in this scenario. But there's this which automates the re-upload into trilium:

![image](https://user-images.githubusercontent.com/617641/208297243-debb8d7d-a4c9-430a-9096-b9205f428aa9.png)

> How can I debug my script? Any way I can write into the backend log?

Yes - `api.log()` will write into the backend log:

![image](https://user-images.githubusercontent.com/617641/208297374-470b34e2-7d8c-4004-ae3f-04558b0f0595.png)

**Comment by TheOneValen** [comment]
Hi. `utcDateModified` on a note seems to give me the creation datetime everytime. So does `dateModified`. Am I doing something wrong or is this a bug? @zadam

**Reply by TheOneValen** [reply]
Also: ~runOnNoteChange=... does not seem to trigger my script :(

**Reply by TheOneValen** [reply]
The note is modified by `setContent` calls. (the one with the trigger and the one with the wrong dateModified, die UI ' note info' does show the modified date correctly)

**Reply by zadam** [reply]
Hi, the problem you're running again is that in some ways there are two separate entities "Note" and "NoteContents" which have separate dateModified, but this complexity is hidden on UI. If you want notecontent's datemodified, you can call `note.getContentMetadata()` (on backend API).

Similar problem happens with `runOnNoteChange` - this works only with the note properties and not with content. I've just added a separate event `runOnNoteContentChange` to master for this purpose.

**Reply by TheOneValen** [reply]
Wow thank you for the explanation. I'll refactor that part. Currently I managed to get the bookmark syncing stuff to actually put the data into trilium. It ends up as a json document (with manual execute scripts in between because of the event stuff). Next step is building a note tree out of the json tree :)

**Reply by TheOneValen** [reply]
Hm. So creating 8k documents out of one json document is extremely slow. Trilium hangs for quite a while. When this runs in the background the server instance would be unavailable for a long time. MAybe I have to go another route and fetch the json file by an external script that then syncs the json into the note tree via ETAPI? Is there some best practice for this case?

Seems just creating is slow. Updating is faster....

Is there an attribute to exclude notes from `ctrl+j` and is there a limit on how many notes I can manage with trilium?

also: is there a method to move a note to a different parent? I don't find a `move` method or something like that... DO I have to use `ensureNoteIsAbsentFromParent` with `ensureNoteIsPresentInParent`?

Using those 2 method it works :)

**Reply by zadam** [reply]
Hmm, creating notes shouldn't be that slow. It should do at least like 10 notes per second, but usually much faster. 

> Is there an attribute to exclude notes from ctrl+j

there's an `#archived` label for that. It can also be set as inheritable for common parent.

> and is there a limit on how many notes I can manage with trilium?

Target is 100 000 notes, but could be more. My instance has 30 000 and things are snappy.

> also: is there a method to move a note to a different parent? I don't find a move method or something like that... DO I have to use ensureNoteIsAbsentFromParent with ensureNoteIsPresentInParent?

That's essentially "move". The reason why it's so awkwardly named is because in trilium every note can have multiple parents. Semantics of "move" are then unclear.

**Reply by TheOneValen** [reply]
10 notes/s can be, maybe even more. I wanted to create 10k notes (1 note for 1 bookmark) and hence it looks slow.

I basically re-implemented the server-side api of https://www.xbrowsersync.org/. The browser plugin now stores the encrypted bookmarks in a trilium note. With another scripts (manually run until the `runOnNoteContentChange` event is released) the data gets decrypted and the browser bookmark hierarchy gets created and kept up to date (move/rename/delete). It is only xbs->trilium though. Any changes to the structure and data made will not be synced back and overwritten on the next sync. Only the content of the note stays as modified in trilium. My goal was to make it possible to have notes on websites and/or link them into other notes. The multi-parent/clone feature makes it possible to have also clone them into another part of my notes. Very handy.

You think someone else would be interested in this? Then I can put it into show & tell when its "done". It needs some reverse proxy rewrite rules, though.

Thanks again for your support!

**Reply by zadam** [reply]
> You think someone else would be interested in this?

TBH I don't know, but a "Show & Tell" would be certainly welcome. Could be useful also as a more complex example of scripting.

**Reply by TheOneValen** [reply]
> Hi, the problem you're running again is that in some ways there are two separate entities "Note" and "NoteContents" which have separate dateModified, but this complexity is hidden on UI. If you want notecontent's datemodified, you can call `note.getContentMetadata()` (on backend API).
> 
> Similar problem happens with `runOnNoteChange` - this works only with the note properties and not with content. I've just added a separate event `runOnNoteContentChange` to master for this purpose.

I installed 58.4 but it does not seem to work. Is this correct: I put the relation on the note which content change I want to "watch". The script note does not get triggered. Am I missing something?

**Reply by zadam** [reply]
@TheOneValen yes, sounds correct. I just tested it and it seems to work for me. Just to be sure, is it a JS backend note? All the events work only for backend scripts.

**Reply by TheOneValen** [reply]
yes so it is like that:

js backend `web` script modifies `text` note via a web hook thingy
`text` note has the event relation to another js backend script `decode`
`decode` then decodes `text` to another note `json`
`json` has another event relation to another js backend script `sync`
`sync` reads `json` and does stuff

The modification `web` does works fine, but not event `decode` is triggered.

triggering `decode` then `sync` manually does work, but is very inconvenient.

**Reply by zadam** [reply]
I see, I didn't know the change was triggered from a script and tested on an UI. I could reproduce and fix the issue - the event was fired from a service layer and thus direct change from `note.setContent()` did not fire it. Will be in the next patch release.

**Reply by TheOneValen** [reply]
It works like a charm! Thank you! I'll clean up my stuff and put it into show and tell as soon as i am able :)
